certbot/lock.py

```diff
@@ -0,0 +1,139 @@
+"""Implements file locks for locking files and directories in UNIX."""
+import errno
+import fcntl
+import logging
+import os
+
+from certbot import errors
+
+logger = logging.getLogger(__name__)
+
+
+def lock_dir(dir_path):
+    """Place a lock file on the directory at dir_path.
+
+    The lock file is placed in the root of dir_path with the name
+    .certbot.lock.
+
+    :param str dir_path: path to directory
+
+    :returns: the locked LockFile object
+    :rtype: LockFile
+
+    :raises errors.LockError: if unable to acquire the lock
+
+    """
+    return LockFile(os.path.join(dir_path, '.certbot.lock'))
+
+
+class LockFile(object):
+    """A UNIX lock file.
+
+    This lock file is released when the locked file is closed or the
+    process exits. It cannot be used to provide synchronization between
+    threads. It is based on the lock_file package by Martin Horcicka.
+
+    """
+    def __init__(self, path):
+        """Initialize and acquire the lock file.
+
+        :param str path: path to the file to lock
+
+        :raises errors.LockError: if unable to acquire the lock
+
+        """
+        super(LockFile, self).__init__()
+        self._path = path
+        self._fd = None
+
+        self.acquire()
+
+    def acquire(self):
+        """Acquire the lock file.
+
+        :raises errors.LockError: if lock is already held
+        :raises OSError: if unable to open or stat the lock file
+
+        """
+        while self._fd is None:
+            # Open the file
+            fd = os.open(self._path, os.O_CREAT | os.O_WRONLY, 0o600)
+            try:
+                self._try_lock(fd)
+                if self._lock_success(fd):
+                    self._fd = fd
+            finally:
+                # Close the file if it is not the required one
+                if self._fd is None:
+                    os.close(fd)
```

Comment:
I'll add this if you want but I'm not convinced it's necessary. If we try to lock the file, but another process holds the lock, we error out immediately. The only time we repeat this loop is if another process deletes the lock file after we've opened the file but before we've locked it. We only delete the lockfile when releasing an acquired lock so it's very rare for us to loop at all here.

With that said, if you'd still like me to add something like this, let me know and I will.